### PR TITLE
adding a command to gurgler to delete old artifacts

### PR DESCRIPTION
**Pull request by jachin**
This will delete artifacts that are older than 90 days and are _not_ being used in any of the known environments.

### CODE REVIEW

**Review by BenSandeen — APPROVED** [review]
I don't know enough JS to know if the code is doing what you say it is, but I wholeheartedly support the intent

File: bin/gurgler.js [hunk 1]
```diff
@@ -196,4 +196,23 @@ program
     )
   });
 
+
+
+program
+  .command("delete-old-deploys")
+  .description("delete artifacts from S3 that are older than a year and not being used")
```
Comment (sbone) [inline bin/gurgler.js:203]:
```suggestion
  .description("delete artifacts from S3 that are older than 90 days and not being used")
```
If I'm following...

File: bin/v2.js [hunk 2]
```diff
@@ -525,8 +555,111 @@ const releaseCmd = (cmdObj, bucketNames, lambdaFunctions, environments, bucketPa
...
 
+const cleanupCmd = async (cmdObj, bucketNames, lambdaFunctions, environments, bucketPath, packageName) => {
+
+  const now = new Date();
+  const ninetyDaysAgo = new Date(now.getTime() - (1000 * 60 * 60 * 24 * 90));
+
+
+  // This line might be too cleaver. It's getting all the server environments and then getting an array of just the unique ones.
+  // noinspection JSUnresolvedVariable
+  const serverEnvironments = [...new Set(environments.map(environment => environment.serverEnvironment))];
```
Comment (sbone) [inline bin/v2.js:566]: 👀

File: bin/v2.js [hunk 3]
```diff
@@ -525,8 +555,111 @@ const releaseCmd = (cmdObj, bucketNames, lambdaFunctions, environments, bucketPa
     .catch(err => console.error(err));
 }
 
+const cleanupCmd = async (cmdObj, bucketNames, lambdaFunctions, environments, bucketPath, packageName) => {
+
+  const now = new Date();
+  const ninetyDaysAgo = new Date(now.getTime() - (1000 * 60 * 60 * 24 * 90));
+
+
+  // This line might be too cleaver. It's getting all the server environments and then getting an array of just the unique ones.
```
Comment (sbone) [inline bin/v2.js:564]:
```suggestion
  // This line might be too clever. It's getting all the server environments and then getting an array of just the unique ones.
```

File: bin/v2.js [hunk 4]
```diff
@@ -525,8 +555,111 @@ const releaseCmd = (cmdObj, bucketNames, lambdaFunctions, environments, bucketPa
...
+                // only allow an asset to be deleted if it's not being used in any of the environments
+                for (const releasedVersion of releasedVersions) {
+                  // noinspection JSUnresolvedVariable
+                  if (releasedVersion.releasedHash === artifact.hash) {
+                    return false;
+                  }
+                }
+                return true;
+              })
+            console.log(`We are going to delete ${oldArtifacts.length} ${packageName} artifacts in ${serverEnvironment}`);
```
Comment (sbone) [inline bin/v2.js:604]:
```suggestion
            console.log(`We are going to delete ${oldArtifacts.length} ${packageName} artifact(s) in ${serverEnvironment}`);
```

For that day this only deletes one artifact 😄

**Review by sbone — APPROVED** [review]
Nice! A couple comment suggestions.

File: bin/v2.js (hunk 2 above)
Comment (BenSandeen) [reply bin/v2.js:566]: I don't know what the `...` part does, but getting a collection of unique items by converting things to a set is not especially unusual

File: bin/v2.js [hunk 5]
```diff
@@ -207,22 +248,17 @@ const getDeployedVersionList = (bucketName, bucketPath) => {
...
+  _.reverse(_.sortBy(versions, ['lastModified']))
     .slice(0, size).forEach(version => {
 
     const {base, ext} = path.parse(version.filepath);
     const split = base.split(".");
 
-    // This check is technically not necessary assuming getDeployedVersionList only returns versions
-    // matching this criteria.
+    // This check is strictly not necessary assuming getDeployedVersionList only returns versions
+    //  matching these criteria.
```
Comment (BenSandeen) [inline bin/v2.js:258]: Since we're doing comment suggestions, I'll add my own
```suggestion
    // matching these criteria.
```

File: bin/v2.js [hunk 6]
```diff
@@ -525,8 +555,111 @@ const releaseCmd = (cmdObj, bucketNames, lambdaFunctions, environments, bucketPa
     .catch(err => console.error(err));
 }
 
+const cleanupCmd = async (cmdObj, bucketNames, lambdaFunctions, environments, bucketPath, packageName) => {
+
```
Comment (BenSandeen) [inline bin/v2.js:559]: I'd prefer no newline, but I don't do much JS at all, so take that with a grain of salt

File: bin/v2.js [hunk 7]
```diff
@@ -525,8 +555,111 @@ const releaseCmd = (cmdObj, bucketNames, lambdaFunctions, environments, bucketPa
...
+  const now = new Date();
+  const ninetyDaysAgo = new Date(now.getTime() - (1000 * 60 * 60 * 24 * 90));
+
+
```
Comment (BenSandeen) [inline bin/v2.js:563]: I'd prefer no newline, but I don't do much JS at all, so take that with a grain of salt

File: bin/v2.js (hunk 6 above)
Comment (BenSandeen) [reply bin/v2.js:559]:
```suggestion
```

File: bin/v2.js [hunk 8]
```diff
@@ -525,8 +555,111 @@ const releaseCmd = (cmdObj, bucketNames, lambdaFunctions, environments, bucketPa
...
+                  if (err) console.error(err, err.stack); // an error occurred
+                  else console.log(data);           // successful response
+                });
+              }
+              console.log("Deleted.")
+            } else {
+              console.log("Very well, not deleting anything then.")
+            }
+          })
+
```
Comment (BenSandeen) [inline bin/v2.js:655]: I'd prefer no newline, but I don't do much JS at all, so take that with a grain of salt
```suggestion
```

**Pull request merged**